src/kernel/thread.c

```diff
@@ -619,16 +619,16 @@ bool_t ensureSchedulable(tcb_t *thread, bool_t can_timeout)
     sched_context_t *sc = thread->tcbSchedContext;
     if (unlikely(sc == NULL)) {
         /* No associated scheduling context */
-        if (false) {
-            // TODO: Add not bound fault
-            // SCInvalid can only be sent to active threads
+        if (validInvalidSchedContextHandler(thread)) {
+            current_fault = seL4_Fault_InvalidSchedContext_new(seL4_SchedContextMissing);
+            handleTimeout(thread);
         }
         return false;
     } else if (unlikely(!sc_active(sc))) {
         /* Not configured */
-        if (false) {
-            // TODO: Add not bound fault
-            // SCInvalid can only be sent to active threads
+        if (validInvalidSchedContextHandler(thread)) {
```

Comment:
where is ensureSchedulable called? I don't recall this function. When can you fault for not having an SC?

Reply:
`ensureSchedulable` is called basically before any call to `SCHED_ENQUEUE` (including via `possibleSwitchTo`) or `SCHED_APPEND` for a thread that is not the current thread or when the current thread may have changed or lost it's SC.

Reply:
Only exception is in `sendFaultIPC` which, to avoid recursion, only checks `isSchedulable` and will not fault an unblocked receiver for temporal reasons (although `sendFaultIPC` should not be called if there is a blocked receiver that would fault).

Reply:
I did this initially and decided it was far too heavy weight, and better to work on a case-by-case basis. There are plenty of places where SCHED_ENQUEUE is called and we *know* we have a scheduling context.

Reply:
I think all of those places made the assumption that threads blocked on IPC always had a ready and sufficient scheduling context which prevents configuring scheduling contexts without knowing what the IPC state of a thread is. I was under the impression we wanted to keep IPC state orthogonal to SC configuration.

Reply:
We have to keep them separate, since IPC state is mutable.

Reply:
There are a handful of cases where the check isn't added where we were already doing all necessary checking (such as `yieldTo`) but I think they are in the minority.

I think to be more specific, this check is added when we try to reschedule a thread but have no prior knowledge of the SC with which it would be scheduled. This excludes, for example, the fastpath as that will always use the SC that was current at kernel entry.